proxygen/lib/http/codec/compress/experimental/qcram/QCRAMHeaderTable.cpp

```diff
@@ -0,0 +1,365 @@
+/*
+ *  Copyright (c) 2017, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+#include <proxygen/lib/http/codec/compress/experimental/qcram/QCRAMContext.h>
+#include <proxygen/lib/http/codec/compress/experimental/qcram/QCRAMHeaderTable.h>
+#include <folly/experimental/Bits.h>
+
+#include <glog/logging.h>
+
+using std::list;
+using std::pair;
+using std::string;
+
+namespace proxygen {
+
+
+size_t QCRAMNewTableImpl::size() const {
+  return vec_.size();
+}
+
+QCRAMHeader& QCRAMNewTableImpl::operator[] (size_t i) {
+  return vec_[i];
+}
+
+void QCRAMNewTableImpl::init(size_t vecSize) {
+  vec_.reserve(vecSize);
+  for (uint32_t i = 0; i < vecSize; i++) {
+    vec_.emplace_back();
+  }
+}
+
+void QCRAMNewTableImpl::resize(size_t size) {
+  vec_.resize(size);
+}
+
+void QCRAMNewTableImpl::moveItems(size_t oldTail, size_t oldLength, size_t newLength) {
+  std::move_backward(vec_.begin() + oldTail, vec_.begin() + oldLength,
+                     vec_.begin() + newLength);
+}
+
+void QCRAMNewTableImpl::add(size_t head, const HPACKHeaderName& name,
+                            const folly::fbstring& value, int32_t epoch, int32_t packetEpoch) {
+  VLOG(1) << " add vec[" << head << "] name=" << name << " value=" << value;
+  vec_[head].name = name;
+  vec_[head].value = value;
+  vec_[head].epoch = epoch;
+  vec_[head].packetEpoch = packetEpoch;
+}
+
+bool QCRAMNewTableImpl::isVulnerableEpoch(uint32_t i,
+                                          int32_t commitEpoch,
+                  int32_t packetEpoch) const {
+  const QCRAMHeader& header = vec_[i];
+  int32_t epoch = header.epoch;
+  //  return !((epoch <= commitEpoch) || (epoch == packetEpoch));
+  return epoch > commitEpoch;
+}
+
+
+void QCRAMHeaderTable::init(uint32_t capacityVal) {
+  bytes_ = 0;
+  size_ = 0;
+  head_ = 0;
+  capacity_ = capacityVal;
+  uint32_t initLength = getMaxTableLength(capacity_) / 2;
+  table_->init(initLength);
+  names_.clear();
+  VLOG(1) << "table init to capacity " << capacity_;
+}
+
+bool QCRAMHeaderTable::add(const HPACKHeader& header) {
+  return add(header, -1, -1);
+}
+
+bool QCRAMHeaderTable::add(const HPACKHeader& header, int32_t epoch,
+                           int32_t packetEpoch) {
+  if (header.bytes() > capacity_) {
+    // Per the RFC spec https://tools.ietf.org/html/rfc7541#page-11, we must
+    // flush the underlying table if a request is made for a header that is
+    // larger than the current table capacity
+    reset();
+    return false;
+  }
+
+  // Make the necessary room in the table if appropriate per RFC spec
+  if ((bytes_ + header.bytes()) > capacity_) {
+    if (!canEvict(header.bytes(), capacity_)) {
```

Comment:
I feel like it may be worthwhile to evict as much as can be right now, to prevent additional references from being added later.